For comments only. (#3)
lsst-dm/dmtn-137: body.tex

Popular in large-scale scientific computing, it provides command line and API interfaces for scientists to write abstract workflow independent of the underlying computing infrastructure.
Pegasus workflows are expressed as Directed Acyclic Graphs (DAGs) where jobs are the nodes and job dependencies are represented by the edges.
Pegasus utilizes HTCondor DAGMan as its execution engine, but also supports various execution environments and data management strategies.
\subsection{LSST Software Stack}
The \href{https://pipelines.lsst.io/}{LSST Science Pipelines Software Stack} is a collection of data processing codes for optical and near-infrared astronomy, and is used for DRP \citep{2018PASJ...70S...5B}.
Besides image processing and astronomical catalog analysis, it includes a middleware layer for data access, task definition and execution.
The LSST Data Butler is a framework that abstracts the underlying data persistence and implementations from Science Pipelines algorithms, and a data repository is a set of datasets under control of the Data Butler \citep{2018arXiv181208085J}.
The Data Butler concepts can be found in \citeds{LDM-463}.

> Recently, the middleware layer is ungergoing a major refactoring, with the new middleware known as the Generation 3 Middleware ("Gen3" Middleware; \citeds{DMTN-056}).

Reviewer: "The middleware layer is in the course of a major redesign and refactoring, ..."

Reviewer 2: Rephrased. Thanks.

\subsection{Alternative architecture designs}

We discussed different architecture designs but did not pursue all of them due to time constraints.

One prominent idea was to use HTCondor utilities to transfer files with S3, rather than relying on Data Butler to communicate directly with S3. In this design HTCondor controls all file transfers, and all datasets are declared in the execution workflow. The proposed implementation was partially external to the LSST code base, and such a data transfer plug-in was added to the HTCondor code base recently. In each job there is a POSIX data repository for local data access, and a mechanism is needed to trigger and perform data ingestion at a central S3 data repository. This requires additional Generation 3 Middleware utilities for posix-S3 datastore copying and URL generation. These issues disfavored this architecture. A favorable benefit of this approach is the potential of bulk data transfer management optimizations.

With respect to the database, we discussed the possibility of having local registries within individual jobs.
Each job has a fast local SQLite registry.

> After each job finishes, the indiviual registry is then collated into the central database.

Reviewer: "individual"

Reviewer 2: Fixed. Thanks!

Besides simplifying failure recovery, another benefit is to avoid potential bottlenecks related to large scale parallel SQL transactions by bundling multiple transactions together and handling database connections in a centralized manner.
This envisions a fully shared-nothing architecture, but implementing this requires more thorough design work in conjunction with the Generation 3 Middleware development.

> Also, the current implementations of the Butler registry depend on conflict resolution syntax to resolve insertions of complexs datasets, which may not work in this scenario.

Reviewer: "... complex datasets; that may not ..."

Reviewer 2: I'm resolving this and have moved the discussion down to KT's comment that removed this sentence. Thanks!

\section{Approaches and strategies}

\poc's goal is to demonstrate that running Data Release Production (DRP) on AWS is possible.
The following approaches were taken.

\subsection{Progress in phases}

\citeds{DMTN-114} proposed development phases that broadly outlined how progress can be made and when performance would be monitored. The outlined plan was to move the execution to AWS without any backend modifications first and then gradually switch out each component to AWS.

> Before this \poc, the DRP execution relied a shared POSIX-compliant filesystem.

Reviewer: "relied on"

Reviewer 2: Fixed but left open, resolved.

> In the first phase such filesystem was replaced by Amazon EFS.

Reviewer: "this filesystem"

Reviewer 2: Rephrased. Thanks.

> The data repository, including a POSIX datastore and a SQLite registry, was entirely located on EFS that was mouted to each of the EC2 compute instances.

Reviewer: "... on EFS, which was mounted to ..."

Reviewer 2: Fixed thanks!

> Later, the S3 datastore replaced the need of EFS, and the PostgreSQL RDS replaced the SQLite registry.

Reviewer: "Subsequently, support for the S3 datastore eliminated the need for EFS, ..."

Reviewer 2: Was fixed but left open, resolved.

Also, a Personal HTCondor Pool on an single EC2 instance was used before we tested the capability to launch new EC2 instances as the HTCondor workers and execute jobs.
This approach allowed us to debug and adapt more smoothly and always had a fallback option in integrating and testing new features and components.

\subsection{Test minimally before scaling up}

Small tasks are tested before workflows of jobs, and small workflows are tested before large workflows.
The canonical \href{https://github.com/lsst/ci\_hsc}{ci\_hsc} dataset and the accompanying CI workflow provide a minimal HSC test dataset and a representative DRP workflow and algorithms.
This test workflow is an important step in integrating new changes.
For larger scale tests we used the \href{https://jira.lsstcorp.org/browse/DM-11345}{HSC-RC2 dataset}.

\subsection{Use Generation 3 Middleware}

> One of the mandates in the \poc~was to use the Generation 3 Middleware (\citeds{DMTN-056}), which has been designed to ease the DRP execution and automation compared to the previous Generation 2 Middleware. Generation 3 Middleware design is more pliable to alternative backends, compared to Generation 2 Middleware.

Reviewer: "... design is more readily extensible to alternative ..."

Reviewer 2: Was fixed but left open. This made me re-write a sentence in the "Building AWS support into the Data Butler" section that also used pliable. Thanks!

> As it was still relatively early in its implementation, findings in the \poc~were feedbacked to the Generation 3 Middleware team and influenced its development.

Reviewer: "fed back"

Reviewer 2: Was fixed but left open, resolved.

> On the other hand, as during the \poc~the Generation 3 Middleware has been under active development, backwards incompatible changes occurred often, and the APIs were unstable.

Reviewer: "has been" => "was" (if it's going to be qualified by "during the POC").

Reviewer 2: Fixed, thanks!

Furthermore, there were no integration tests when the \poc~started and the test coverage has been suboptimal.
Hence, changes often broke functionality.
Lack of a good integration test is still a major concern going forward.
New integration tests may not be accepted due to the much longer time required to run the test suite.

> Keeping this in mind, We decided not to always follow the bleeding edge version and updated only when unnecessary, such as important bug fixes. The stack versions should not have strong impacts in the \poc~conclusions in terms of the cloud feasibility.

Reviewer: "we" (lowercase)
"... when necessary, such as for ..."
"We believe that the specific stack versions used should not have strong impacts on the \poc~conclusions regarding the feasibility of cloud deployment of DRP."

Reviewer 2: Was fixed but left open, resolved.

EC2 instances come with multiple instance types for a configuration of memory, CPU, and storage.
An Amazon Machine Image (AMI) bundles the Operating System (OS) with required data and configurations into a convenient image that can be launched on any EC2 instance type.
\href{https://aws.amazon.com/efs/}{Amazon Elastic File System (Amazon EFS)} provides a POSIX-compliant, scalable, elastic NFS file system designed to provide parallel shared access to thousands of AWS compute instances.
We have used On-demand and Spot instances in the \poc.
On demand instances cost more but they can not be deallocated by AWS. Spot instances are significantly cheaper, but at any time AWS, given a 2 mintue warning, can request the compute resources back and deallocate them.
\subsection{Amazon Simple Storage Service (S3)}
\href{https://aws.amazon.com/s3/}{Simple Storage Service (S3)} is an object storage that allows massive amounts of unstructured data where each object typically includes the data, related metadata and is identified by a globally unique identifier, to be stored and accessed from EC2 instances or elsewhere in a durable and highly scalable way.

> Buckets, organizational units in which related objects are generally stored, enable easier access and priviledges administration. Access, read, write, delete and other atomical units of action on the objects themselves can be allowed or forbidden at the account, bucket or individual object level. Logging is available for all actions on the Bucket level and/or at the individual object granularity. It is also possible to define and issue complex alert conditions on Bucket or object actions which can execute arbitrary actions or workflows, but S3 itself is not POSIX compliant.

Reviewer 2: I do not agree that it would be necessary to re-vamp the workflow system to take advantage of actions and triggers. Using Lambda we can create this functionality completely external to the workflow system. More than just invoking additional processing however, users can trigger on specific file types and names - f.e. pegasus performance logs, tags, etc. - to create for themselves notifications, or automate graph/table creation that they use completely unrelated to the actual data processing. 
I have left the sentence as suggested - since most of it seems to be spelling mistake fixes.

As for the last point: I've redone the document to be one sentence per line. Didn't notice that recommendation initially. Thanks.

\subsection{PostgreSQL}

> PostgreSQL is one of the most popular open source relational database systems available. The choice to go with PostgreSQL was based on the fact that it's a very popular and well supported open source software that suffers from no additional licensing fees usually associated with proprietary software. \href{https://aws.amazon.com/rds/}{Relational Database Service (RDS)} is the AWS cloud service that launches and configures databases with ease.

Reviewer 2: Fixed but left open, resolved. Thanks!

The Data Butler concepts can be found in \citeds{LDM-463}.
Recently, the middleware layer is in the course of a major redesign and refactoring, with the new middleware known as the Generation 3 Middleware ("Gen3" Middleware; \citeds{DMTN-056}).
This \poc~added new features in the Generation 3 Data Butler; see Sect \ref{sec:butler}.

\section{Architecture design}

The system at the end of the PoC is shown in Figure~\ref{fig:arch}.
All components are hosted on the AWS platform.

\begin{figure}
  \centering
  \includegraphics[width=\textwidth]{figures/arch}
  \label{fig:arch}
  \caption{Diagram of the AWS PoC system architecture.}
\end{figure}

AMIs containing LSST Software Stack, Pegasus, and configured HTCondor are pre-built in two flavors: master and worker. These contain the complete environment required to run DRP workflows and scale them out with HTCondor. The user launches an on-demand EC2 instance using the master AMI to be the submit host and the central manager of HTCondor pool.
The user logs in to this master to control the workflow execution.

The workflow of the LSST DRP in Generation 3 Middleware is represented as a Directed Acyclic Graph (DAG) known as Quantum Graph.

> In the Generation 3 Middleware, input and output datasets are defined in the Pipeline Tasks and each unit is known as a Quantum.

Reviewer 2: I believe this was resolved in the latest commit by Hsin-Fang. Thanks!

The Quantum Graph is aware of the Quantum dependency and determines the execution workflow.
For simplicity we map each Quantum to a standalone job, despite this may not be optimal for the performance.
Pegasus is used to submit the workflow to HTCondor as well as monitor the workflow execution.

> Each LSST Pipeline Task Quantum is resource-independent. Pegasus adds other necessary jobs, such as data transfer, to the executable workflow. HTCondor DAGMan is the workflow execution engine behind Pegasus and controls the processes.

Reviewer 2: Fixed but left open, resolved.

Compute resources are procured from the master node via \texttt{condor\_annex} by targeting the worker AMI. Annex is a resource acquisition and external compute resource separation management unit. Each resource pool can have multiple annexes. Each annex manages its own lifecycle. It is possible to request on-demand or spot instance fleets.

LSST Pipeline jobs are executed on the worker instances; they use Data Butler for data access.
New Data Butler backends were implemented during the PoC, see Sect \ref{sec:butler}.

> The Butler datastore is located in an S3 Bucket and follows the same hierarchical structure that POSIX datastore does. Consumed and produced datasets are read and written directly from S3 as bytes, whenever possible, and only downloaded to temporary files for objects that are not serializable.

Reviewer 3: @DinoBektesevic I got this from some sentences you wrote originally and probably misunderstood. May you please clarify?

> The Butler datastore is located in an S3 Bucket and follows the same hierarchical structure that POSIX datastore does. Consumed and produced datasets are read and written directly from S3 as bytes, whenever possible, and only downloaded to temporary files for objects that are not serializable.

Reviewer 2: I have rephrased to: 

> Consumed and produced datasets are read and written directly from S3 as bytes, whenever possible, and only downloaded to temporary files for objects whose formatters do not support serialization.

 Is this better? 

Not all Formatters support reading/writing to/from bytes. This means that certain Datasets that use those Formatters will always be downloaded to a temporary file.

Since the directory structure is preserved by the S3 datastore the entire data repository is trivially transferable between the cloud and a local filesystem.

> The Butler registry is a RDS PostgreSQL database that keeps track of all LSST science files.

Author: ```suggestion
The Butler registry is an RDS PostgreSQL database that keeps track of all LSST science files.
```

> Files that are not managed by Data Butler are managed by the Condor File IO via Pegasus. These include the Data Butler configuration file, pipeline definition (Quantum) files, and the log files. The master instance serves as the staging site for these files. Various network protocols transfer the files between the master and the workers; instances do not share a filesystem.

Author: What are the network protocols for file transfer?

Reviewer 3: A quick search in the Pegasus documents gave me: GridFTP, SCP, SRM, Amazon S3, Google Storage, XRootD, HTTP, and local cp/symlinking 

But I think they are not relevant here or we meant different Pegasus configurations. I'll rephrase.

\subsection{Alternative architecture designs}

We discussed different architecture designs but did not pursue all of them due to time constraints.

> One prominent idea was to use HTCondor utilities to transfer files with S3, rather than relying on Data Butler to communicate directly with S3. In this design HTCondor controls all file transfers, and all datasets are declared in the execution workflow. The proposed implementation was partially external to the LSST code base, and such a data transfer plug-in was added to the HTCondor code base recently. In each job there is a POSIX data repository for local data access, and a mechanism is needed to trigger and perform data ingestion at a central S3 data repository. This requires additional Generation 3 Middleware utilities for posix-S3 datastore copying and URL generation. These issues disfavored this architecture. A favorable benefit of this approach is the potential of bulk data transfer management optimizations.

Author: ```suggestion
One prominent idea was to use HTCondor utilities to transfer files from and to S3, rather than relying on the Data Butler to communicate directly with S3. In this design HTCondor controls all file transfers, and all datasets are declared in the execution workflow. This design is enabled by a data transfer plug-in that was recently added to HTCondor. In each job there would be a local POSIX data repository for all data access. Utilities would be needed for determining, based on the Quantum Graph, the URLs of datasets that need to be transferred; for setting up the local Butler registry; and for ingesting the HTCondor-transferred S3 datasets into the main Butler registry. This additional work was not able to be scheduled as part of the PoC. A benefit of this approach is the potential for optimizing bulk data transfers.
```

With respect to the database, we discussed the possibility of having local registries within individual jobs.
Each job has a fast local SQLite registry.

> After each job finishes, the individual registry is then collated into the central database.

Author: Add "This is a subset of the previous design alternative, with only the registry being localized and not the datasets."

Reviewer 2: Fixed but left open. Resolved.

Besides simplifying failure recovery, another benefit is to avoid potential bottlenecks related to large scale parallel SQL transactions by bundling multiple transactions together and handling database connections in a centralized manner.
This envisions a fully shared-nothing architecture, but implementing this requires more thorough design work in conjunction with the Generation 3 Middleware development.

> Also, the current implementations of the Butler registry depend on conflict resolution syntax to resolve insertions of complex datasets, that may not work in this scenario.

Author: I don't think this is a problem; at worst it limits the performance, but fundamentally any insertions into the local SQLite can be done in the global registry as well.

\section{Approaches and strategies}

\poc's goal is to demonstrate that running Data Release Production (DRP) on AWS is possible.
The following approaches were taken.

\subsection{Progress in phases}

\citeds{DMTN-114} proposed development phases that broadly outlined how progress can be made and when performance would be monitored. The outlined plan was to move the execution to AWS without any backend modifications first and then gradually switch out each component to AWS.

Before this \poc, the DRP execution relied on a shared POSIX-compliant filesystem.

> In the first phase Amazon EFS is used as this filesystem.

Author: ```suggestion
In the first phase Amazon EFS was used as this filesystem.
```

> The data repository, including a POSIX datastore and a SQLite registry, was entirely located on EFS, which was mouted to each of the EC2 compute instances.

Author: ```suggestion
The data repository, including a POSIX datastore and a SQLite registry, was entirely located on EFS, which was mounted on each of the EC2 compute instances.
```

Subsequently, support for the S3 datastore eliminated the need of EFS, and the PostgreSQL RDS replaced the SQLite registry.
Also, a Personal HTCondor Pool on an single EC2 instance was used before we tested the capability to launch new EC2 instances as the HTCondor workers and execute jobs.

> This approach allowed us to debug and adapt more smoothly and always had a fallback option in integrating and testing new features and components.

Author: ```suggestion
This approach allowed us to debug and adapt more smoothly, ensuring we always had a fallback option while integrating and testing new features and components.
```

\subsection{Test minimally before scaling up}

Small tasks are tested before workflows of jobs, and small workflows are tested before large workflows.

> The canonical \href{https://github.com/lsst/ci\_hsc}{ci\_hsc} dataset and the accompanying CI workflow provide a minimal HSC test dataset and a representative DRP workflow and algorithms.

Author: Mention the data size and number of jobs.

This test workflow is an important step in integrating new changes.

> For larger scale tests we used the \href{https://jira.lsstcorp.org/browse/DM-11345}{HSC-RC2 dataset}.

Author: Again, the data size and number of jobs (actually run) would be good here.

\subsection{Use Generation 3 Middleware}

One of the mandates in the \poc~was to use the Generation 3 Middleware (\citeds{DMTN-056}), which has been designed to ease the DRP execution and automation compared to the previous Generation 2 Middleware. Generation 3 Middleware design is more readily extensible to alternative backends, compared to Generation 2 Middleware.
As it was still relatively early in its implementation, findings in the \poc~were fed back to the Generation 3 Middleware team and influenced its development.

On the other hand, as during the \poc~the Generation 3 Middleware was under active development, backwards incompatible changes occurred often, and the APIs were unstable.

> Furthermore, there were no integration tests when the \poc~started and the test coverage has been suboptimal.

Author: ```suggestion
Furthermore, there were no integration tests when the \poc~started, and unit test coverage was incomplete.
```

Hence, changes often broke functionality.

> Lack of a good integration test is still a major concern going forward.

Author: ```suggestion
Lack of a good integration test for the code written for this \poc is still a major concern going forward.
```

> New integration tests may not be accepted due to the much longer time required to run the test suite.

Author: I would delete this.  It doesn't take any longer if we can run in parallel.  It's a matter of resources.

> Keeping this in mind, we decided not to always follow the bleeding edge version and updated only when unnecessary, such as for important bug fixes.

Author: ```suggestion
Keeping this in mind, we decided not to always follow the bleeding edge version of the middleware, updating only when necessary, such as for important bug fixes.
```

We believe that the specific stack versions used should not have strong impacts on the \poc~conclusions regarding the feasibility of cloud deployment of DRP.

\subsection{Focus on end-to-end execution}

Potential optimizations and further investigations were identified throughout the PoC project but were not carried out.
Some ideas are described in Sec \ref{future}.

\section{Building AWS support into the Data Butler}
\label{sec:butler}

The Data Butler is the overarching data IO abstraction through which all LSST data access is mediated \citeds{DMTN-056}. Datasets are referred to by their unique IDs, or a set of identifying references, which are then resolved through a Registry that matches the dataset IDs, or references, to the location, file format and the Python object type of the dataset. The system that persists, reads and potentially modifies the datasets is called the Datastore. The Registry is almost always backed by a SQL database and the Datastore is usually backed by a shared filesystem. A major focus of the \poc~was to implement, and investigate issues related to, an S3 backed Datastore and a PostgreSQL backed Registry.

> At the time the \poc~project began, Generation 3 Data Butler implemented \texttt{PosixDatastore}, a local or shared filesystem datastore, and a SqliteRegistry. OracleRegistry followed soon after the \poc~work began. Initially the focus was on implementing an S3 backed datastore called S3Datastore. The interface between AWS services and LSST Stack would be based on the official AWS SDK called boto3. In March 2019, Dino Bektesevic visited Tim Jenness for this work and implemented the early versions of a new module in the \texttt{daf\_butler} called \texttt{s3utils}, an \texttt{S3Datastore} class, the \texttt{PosixDatastore} equivalent, and a set of appropriate unit tests that demonstrated its functionality and correctness. The unit tests for the datastore utilize the \texttt{moto} library which mocks requests and responses sent to AWS services, so that no additional external infrastructure is required to use it. \texttt{PostgreSqlRegistry} class was implemented partially during the visit and completed shortly after the visit. The initial implementation showcasing the required changes to the code was submitted as a Draft Pull Request \href{https://github.com/lsst/daf_butler/pull/147}{PR-147} in \texttt{daf\_butler}.

Author: In this section, I think it's a bit excessive to essentially document ticket-by-ticket changes to the code base, along with specific meetings and PRs.  These details could be moved to an appendix or to footnotes (e.g. for the PRs); this section should summarize.

Reviewer 2: I see @hsinfang already pruned this section - thanks for that(!) and sorry for the late response to the PR, but like I said bellow I must've missed the notifications. 

I was asked to document everything I did - so I did. I can prune further if required?

Reviewer 3: (So far I haven't reorganized the session, so I'm leaving this unresolved for now)

Reviewer 3: I moved the Butler section to the Appendix. For now I don't see an individual short Butler section is still needed; instead I expanded the Tech Stack section and the Architecture section where Butler ideas are introduced and explained.

Adding the support for AWS into the Butler exercised a significant fraction of the Generation 3 Data Butler. During the process many faults and unpredictable behaviors were discovered and solved. Many problems touched, and continue to exercise, the general Generation 3 Data Butler implementation, as well as assumptions made during their implementation. Recounting the wide list of major improvements to the codebase, hopefully, reveals how productive this exercise has been in helping generalizing and strengthening the whole Generation 3 Data Butler codebase.

\subsection{PostgreSQL performance}

> The following described performance issue applies strictly to w\_2019\_38 builds and earlier. The QuantumGraph is created by issuing a very large SQL statement that, effectively, creates a cartesian product between many, of the 15-ish or so, tables in the registry. The results of the large query are then parsed in Python and a slew of many different, small, follow-up queries are issued. This diverse workflow presents a difficult challenge in DB optimization. The Generation 3 Middleware team reports that on average it takes 0.5 to 1.5h to create QuantumGraphs for tract-size DRP workflows with Oracle on NCSA infrastructure. In our tests we were unable to create even the simplest QuantumGraphs, even after doubling of the RDS instance resources, and were forced to abort multiple times after 30+h of execution. The lack of performance did not seem related to hardware limitations, since none of the performance metrics availible showed heavy hardware loads nor didthe problem go away even after significant increase of RDS resources.

Author: In this part, aside from fixing the minor spacing and spelling problems, I think it's important to distinguish between the state of the schema and SQL before and after various changes and refactors that occurred, culminating in Jim's latest updates.  Even with the caveat in the first sentence, use of the present tense ("is created", "reports") gives the impression that the description still applies to the current code.

```suggestion
Versions of the Generation 3 Middleware prior to and including w\_2019\_38 did not attempt to optimize database performance. In those versions, the QuantumGraph was created by issuing a very large SQL statement that, effectively, created a cartesian product between many of the tables in the registry. The results of this large query were then parsed in Python, and a slew of many different, small, follow-up queries were issued. The Generation 3 Middleware team reported that on average it took 0.5 to 1.5h to create QuantumGraphs for tract-size DRP workflows with Oracle on NCSA infrastructure using these versions. In our tests we were unable to create even the simplest QuantumGraphs, even after doubling the RDS instance resources, and were forced to abort multiple times after 30+h of execution. None of the performance metrics available showed heavy hardware loads.
```

> PostgreSQL, and Oracle, regularly collect statistics on database objects used in queries. These statistics are then used to generate execution plans. These execution plans are then oftentimes cached. This in general results in performance gains. But because Butler generates all SQL dynamically there are no guarantees that a cached, or even pre-prepared execution plan can be executed. In PostgreSQL many of these optimizations then fail.

Author: I don't think it was demonstrated that cached execution plans had anything to do with the problem.  I would remove this paragraph.

Reviewer 2: Selecting from a view is as fast or as slow as the underlying SQL subquery statement - at least it should be in ideal conditions. To guarantee the best performance recommendation is to include conditions in the view's definition. The optimizer tries to create the optimal execution plan by pushing the predicates on a view into the view before executing the statement but for any/most RDBMS this can sometimes fail for complex queries where conditions are several layers away from the view. 

Defining functions, preparing and caching uniquely optimized execution plans for these large queries helps avoid situations where optimizer fails. We don't use prepared or cached statements - but perhaps they would help. The bonus point is that the way the SQL statements are generated sometimes works against PostgreSQL's internals and similar situations, where we expect the RDBMS to just work without providing it with sufficient information, will continue to occur. This is a consequence of the current registry design.

> Specifically what seems to happen with PostgreSQL is that all the views materialize completely on disk. Even when the views are part of larger statements, in which the outter statements have strict constraints on them, the outter constraints do not seem to penetrate to the view statement. It is unclear, from within \poc, whether Oracle DBMS does not suffer from this issue because it has a better SQL statement optimization engine or if this is something more directly tied to how PostgreSQL implements views. 

Author: Fix "outer", but also leave out the Oracle statement: it doesn't add any information.

Reviewer 2: Fixed "outter". Removed last sentence. Resolved.

> The performance penalty of the behaviour is debilitating. Bellow is an abreviated query plan for the simplest case of \lstinline[basicstyle=\ttfamily]{select * from visit_detector_patch_join limit 4;} that suffers from the described issue.

Author: ```suggestion
This behavior causes the severe performance problem. Below is an abbreviated query plan for the simplest case of \lstinline[basicstyle=\ttfamily]{select * from visit_detector_patch_join limit 4;} that suffers from the described issue.
```

\begin{lstlisting}[style=sqlprompt]
Limit .... (actual time=12732.766..12732.774 rows=4 loops=1)
->Unique .... (actual time=12732.765..12732.770 rows=4 loops=1)
  ->Sort .... (actual time=12732.764..12732.767 rows=6 loops=1)
    Sort Key: ....
    Sort Method: external merge  Disk: 209008kB
    ->Hash Join .... (actual time=763.846..1524.504 rows=4642107 loops=1)
      Hash Cond: ....
      ->Seq Scan on .... (actual time=0.008..16.289 rows=206960 loops=1)
      ->Hash .... (actual time=763.723..763.724 rows=3259107 loops=1)
        Buckets: 65536  Batches: 64  Memory Usage: 3635kB
        ->  Seq Scan on patch_skypix_join .... (actual time=0.006..240.128 rows=3259107 loops=1)
Planning time: 0.323 ms
Execution time: 12759.391 ms
\end{lstlisting}

> Note that the total materialized size of the view on disk is 200MB and that it took 12 seconds to retrieve only 4 results! There is no clear solution to this problem from within the Middleware codebase. 

Author: ```suggestion
Note that the total materialized size of the view on disk is 200MB and that it took 12 seconds to retrieve only 4 results! We were unable to find a way to disable the view materialization in PostgreSQL without changing the SQL queries generated by the middleware.
```

> The sollution is however possible by manually redefining the views as materialized views instead, adding triggers that recreate the views on any insert statement to the underlying tables that make the view, and then adding indexing onto the materialized views. This reduces the time it takes to create a QuantumGraph to the point where it's comparable to that reported by the DM team. The query plan for the same SQL select statement as described above now looks like:

Author: ```suggestion
One solution we found was to manually redefine the views as materialized views instead, adding triggers that recreate the views on any insert statement to the underlying tables that make the view, and then adding indexing onto the materialized views. This reduces the time it takes to create a QuantumGraph to the point where it's comparable to that reported by the DM team. The query plan for the same SQL select statement as described above now looks like:
```

\begin{lstlisting}[style=sqlprompt]
Limit .... (actual time=0.009..0.010 rows=4 loops=1)
->Seq Scan on visit_detector_patch_join .... (actual time=0.008..0.008 rows=4 loops=1)
Planning time: 0.053 ms
Execution time: 0.021 ms
\end{lstlisting}

> Recently \jira{DM-17023} reworked and completely re-implemented the SQL schema and data model for the registries. We have not yet tested the performance of the new registry schema in PostgreSQL. The issues described here well optimize the set of specific queries, at the cost of a performance hit for all inserts into visits, patches..., but an important take away lesson here should be that because all of the SQL is generated completely dynamically guaranteeing execution plan stability and performance is dificult. Not all DBMSs optimize SQL statements equally well. Tying the CI and testing to only one database doesn't suficiently test the code performance and because the SQL is generated completely dynamically to fix potential poor performance requires schema modifications to be made external to the Middleware code. Because some of the performance loss can be debilitating, it is possible that certain DBMSs are, in this implementation, tied into implementing unoptimal solutions just to be able to run DRP.

Author: The dynamic SQL is always of a particular form.  I don't think we have shown that there are _data-dependent_ performance problems.  RDBMS optimization difficulties and differences between RDBMSes apply to static SQL as well as dynamic SQL.  I would rewrite this paragraph.

Reviewer 2: Not sure the comment relates, or maybe I don't understand it. The claim is that because the SQL statements are created dynamically, changes in the code-base affects the statements. A lot of work and effort (multiple weeks) was exerted in optimizing the queries in w38 build. That work is inconsequential for w39 build. Granted in w39 a lot changed, but that doesn't need to be the case. If someone changed a single line of code, that changes the order of select statements or order of conditions in it, just that could invalidate the work that went into getting it to work. These changes could come even outside from the middleware - by lets say SQLAlchemy changing the internals of how they produce the SQL statements. 

Dynamically created statements are perpetually the same, I agree. I don't know how data changes could affect the registry drastically (since most of the data in it is a simple int or string anyhow). The performance is very dependent on how that code produces the statements though and it takes a lot of effort to optimize it. It could also sometimes be impossible in itself, or at least will bring penalties on some other front (f.e. materialized views have triggers on visit inserts).

Reviewer 3: I rewrote this paragraph based on what I understand from here.
